File: stuff_usecases/src/test/java/android/montreal/devfest/stuff/usecases/ManageRandomStuffUseCaseImplTest.java
package android.montreal.devfest.stuff.usecases;

import android.montreal.devfest.stuff.entities.RandomStuffEntity;
import android.montreal.devfest.stuff.policies.RandomStuffVerifier;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class ManageRandomStuffUseCaseImplTest {
    @Mock
    private RandomStuffVerifier verifier;

    @Mock
    private RandomStuffSetter setter;

    @Mock
    private RandomStuffGetter getter;

    private ManageRandomStuffUseCaseImpl useCase;

    @Before
    public void setup() {
        useCase = new ManageRandomStuffUseCaseImpl(verifier, setter, getter);
    }

    @Test
    public void setRandomStuff_whenStuffNotVerified_shouldThrowAnException() {
        // Given
        final RandomStuffData data = new RandomStuffData(
                UUID.randomUUID().toString(),
                UUID.randomUUID().hashCode()
        );

        doReturn(false)
                .when(verifier)
                .isValid(any(RandomStuffEntity.class));

        // Then When
        assertThatExceptionOfType(SetRandomStuffException.class)
                .isThrownBy(() -> useCase.setRandomStuff(data));

        verify(verifier)
                .isValid(any(RandomStuffEntity.class));
    }

    @Test
    public void setRandomStuff_whenStuffVerified_shouldProcessWithoutError() throws Exception {
        // Given
        final RandomStuffData data = new RandomStuffData(
                UUID.randomUUID().toString(),
                UUID.randomUUID().hashCode()
        );

        doReturn(true)
                .when(verifier)
                .isValid(any(RandomStuffEntity.class));
        doNothing()
                .when(setter)
                .setRandomStuff(data);

        // When
        useCase.setRandomStuff(data);

        // Then
        verify(verifier)
                .isValid(any(RandomStuffEntity.class));
        verify(setter)
                .setRandomStuff(data);
    }
}

File: stuff_usecases/src/test/java/android/montreal/devfest/stuff/usecases/ManageRandomStuffUseCaseImplTest.kt
package android.montreal.devfest.stuff.usecases

import android.montreal.devfest.stuff.policies.RandomStuffVerifier
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner
import java.util.*

@RunWith(MockitoJUnitRunner::class)
class ManageRandomStuffUseCaseImplTest {
    @Mock
    private lateinit var verifier: RandomStuffVerifier

    @Mock
    private lateinit var setter: RandomStuffSetter

    @Mock
    private lateinit var getter: RandomStuffGetter

    private lateinit var useCase: ManageRandomStuffUseCaseImpl

    @Before
    fun setup() {
        useCase = ManageRandomStuffUseCaseImpl(verifier, setter, getter)
    }

    @Test
    fun setRandomStuff_whenStuffNotVerified_shouldThrowAnException() {
        // Given
        val data = RandomStuffData(
                UUID.randomUUID().toString(),
                UUID.randomUUID().hashCode()
        )

        doReturn(false)
                .`when`(verifier)
                .isValid(any())

        // Then When
        assertThatExceptionOfType(SetRandomStuffException::class.java)
                .isThrownBy { useCase.setRandomStuff(data) }

        verify(verifier).isValid(any())
    }

    @Test
    @Throws(Exception::class)
    fun setRandomStuff_whenStuffVerified_shouldProcessWithoutError() {
        // Given
        val data = RandomStuffData(
                UUID.randomUUID().toString(),
                UUID.randomUUID().hashCode()
        )

        doReturn(true)
                .`when`(verifier)
                .isValid(any())
        doNothing()
                .`when`(setter)
                .setRandomStuff(data)

        // When
        useCase.setRandomStuff(data)

        // Then
        verify(verifier).isValid(any())
        verify(setter).setRandomStuff(data)
    }

    private fun <T> any(): T {
        Mockito.any<T>()
        return uninitialized()
    }

    private fun <T> uninitialized(): T = null as T
}
